executor/analyze.go

```diff
@@ -408,6 +410,41 @@ func (e *AnalyzeExec) analyzeWorker(taskCh <-chan *analyzeTask, resultsCh chan<-
 	}
 }
 
+func analyzeColumnsPushdownWithRetry(e *AnalyzeColumnsExec) *statistics.AnalyzeResults {
+	analyzeResult := analyzeColumnsPushdown(e)
+	// do not retry if succeed / not oom error / not auto-analyze / samplerate not set
+	if analyzeResult.Err == nil || analyzeResult.Err != errAnalyzeOOM ||
+		!e.ctx.GetSessionVars().InRestrictedSQL ||
+		e.analyzePB.ColReq == nil || *e.analyzePB.ColReq.SampleRate <= 0 {
+		return analyzeResult
+	}
+	finishJobWithLog(e.ctx, analyzeResult.Job, analyzeResult.Err)
+	statsHandle := domain.GetDomain(e.ctx).StatsHandle()
+	if statsHandle == nil {
+		return analyzeResult
+	}
+	var statsTbl *statistics.Table
+	tid := e.tableID.GetStatisticsID()
+	if tid == e.tableInfo.ID {
+		statsTbl = statsHandle.GetTableStats(e.tableInfo)
+	} else {
+		statsTbl = statsHandle.GetPartitionStats(e.tableInfo, tid)
+	}
+	if statsTbl == nil || statsTbl.Count <= 0 {
+		return analyzeResult
+	}
+	newSampleRate := math.Min(1, float64(config.DefRowsForSampleRate)/float64(statsTbl.Count))
+	if newSampleRate >= *e.analyzePB.ColReq.SampleRate {
+		return analyzeResult
+	}
```

Comment:
What if 110000 samples still causes oom? Do we need to handle that case?

Reply:
If 110000 samples still cause oom, it could be a problem with data or quota settings. As we once discussed before, I think we need metrics and alerts for this kind of case to involve ops. Let's do it in another PR.